atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1765,18 +1776,42 @@ protected boolean shouldDeleteAndRollback() {
 
     @Override
     public final void delete(TableReference tableRef, Set<Cell> cells) {
+        deleteInternal(tableRef, cells, ImmutableMap.of());
+    }
+
+    @Override
+    public void deleteWithMetadata(TableReference tableRef, Map<Cell, ChangeMetadata> cellsWithMetadata) {
+        deleteInternal(tableRef, cellsWithMetadata.keySet(), cellsWithMetadata);
+    }
+
+    private void deleteInternal(TableReference tableRef, Set<Cell> cells, Map<Cell, ChangeMetadata> metadata) {
         getCache().delete(tableRef, cells);
-        putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY));
+        putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY), metadata);
     }
 
     @Override
     public void put(TableReference tableRef, Map<Cell, byte[]> values) {
+        putWithMetadataInternal(tableRef, values, ImmutableMap.of());
```

Comment:
I think the cleaner way to do this, is we should just do `putWithMetadata` instead of the internal call. And then we should rename `internalPut` to `internalPutWithMetadata` to make it a bit more clear.

Reply:
I would avoid calling `putWithMedat` from `put` because that would involve a lot of unnecessary byte array copying